src/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs

```diff
@@ -123,13 +123,110 @@ private static Exception CreateNamedMapsNotSupportedException()
+        private static FileAccess TranslateProtectionsToFileAccess(Interop.Sys.MemoryMappedProtections protections)
         {
             return
-                (protections & (Interop.libc.MemoryMappedProtections.PROT_READ | Interop.libc.MemoryMappedProtections.PROT_WRITE)) != 0 ? FileAccess.ReadWrite :
-                (protections & (Interop.libc.MemoryMappedProtections.PROT_WRITE)) != 0 ? FileAccess.Write :
+                (protections & (Interop.Sys.MemoryMappedProtections.PROT_READ | Interop.Sys.MemoryMappedProtections.PROT_WRITE)) != 0 ? FileAccess.ReadWrite :
+                (protections & (Interop.Sys.MemoryMappedProtections.PROT_WRITE)) != 0 ? FileAccess.Write :
                 FileAccess.Read;
         }
 
+        private static FileStream CreateSharedBackingObject(Interop.Sys.MemoryMappedProtections protections, long capacity)
+        {
+            return CreateSharedBackingObjectUsingMemory(protections, capacity)
+                ?? CreateSharedBackingObjectUsingFile(protections, capacity);
+        }
+
+        // -----------------------------
+        // ---- PAL layer ends here ----
+        // -----------------------------
+
+        private static FileStream CreateSharedBackingObjectUsingMemory(
+           Interop.Sys.MemoryMappedProtections protections, long capacity)
+        {
+            // The POSIX shared memory object name must begin with '/'.  After that we just want something short and unique.
+            string mapName = "/corefx_map_" + Guid.NewGuid().ToString("N");
+
+            // Determine the flags to use when creating the shared memory object
+            Interop.Sys.OpenFlags flags = (protections & Interop.Sys.MemoryMappedProtections.PROT_WRITE) != 0 ?
+                Interop.Sys.OpenFlags.O_RDWR :
+                Interop.Sys.OpenFlags.O_RDONLY;
+            flags |= Interop.Sys.OpenFlags.O_CREAT | Interop.Sys.OpenFlags.O_EXCL; // CreateNew
+
+            // Determine the permissions with which to create the file
+            Interop.Sys.Permissions perms = default(Interop.Sys.Permissions);
+            if ((protections & Interop.Sys.MemoryMappedProtections.PROT_READ) != 0)
+                perms |= Interop.Sys.Permissions.S_IRUSR;
+            if ((protections & Interop.Sys.MemoryMappedProtections.PROT_WRITE) != 0)
+                perms |= Interop.Sys.Permissions.S_IWUSR;
+            if ((protections & Interop.Sys.MemoryMappedProtections.PROT_EXEC) != 0)
+                perms |= Interop.Sys.Permissions.S_IXUSR;
+
+            // Create the shared memory object.
+            int fd = Interop.Sys.ShmOpen(mapName, flags, (int)perms);
+            if (fd < 0)
+            {
+                // If ShmOpen fails, fall back to file backing object.
+                // Note that the System.Native shim will force failure on platforms where
+                // the result of native shm_open does not work well with our subsequent call
+                // to mmap.
+                return null; 
```

Comment:
We should only do this fallback if errno is ENOTSUP.  If it's some other failure, we should throw.